packages/imt.sol/contracts/internal/InternalLazyIMT.sol

```diff
@@ -157,6 +157,74 @@ library InternalLazyIMT {
         return _root(self, numberOfLeaves, depth);
     }
 
+    function _merkleProofElements(
+        LazyIMTData storage self,
+        uint40 index,
+        uint8 depth
+    ) internal view returns (uint256[] memory) {
+        uint40 numberOfLeaves = self.numberOfLeaves;
+        require(index < numberOfLeaves, "LazyIMT: leaf must exist");
+        require(depth > 0, "LazyIMT: depth must be > 0");
+        require(depth <= MAX_DEPTH, "LazyIMT: depth must be < MAX_DEPTH");
+
+        uint256[] memory proof = new uint256[](depth);
+        uint256 levelCount = self.numberOfLeaves;
+
+        for (uint8 i = 0; i < depth; ) {
+            // Left leaf
+            if (index & 1 == 0) {
+                bool outsideLimits = ((index + 1) << i) % self.numberOfLeaves != 0;
+                if ((index + 1) < levelCount) {
+                    proof[i] = self.elements[_indexForElement(i, index + 1)];
+                } else if (((index + 1) == levelCount) && (outsideLimits)) {
+                    proof[i] = _root(self, numberOfLeaves, i);
+                } else {
+                    proof[i] = _defaultZero(i);
+                }
+            } else {
+                proof[i] = self.elements[_indexForElement(i, index - 1)];
+            }
+            unchecked {
+                index >>= 1;
+                i++;
+            }
+            levelCount = numberOfLeaves >> i;
+        }
+        return proof;
+    }
+
+    function _merkleProofIndexes(
```

Comment:
I don't think this function should exist in the contract. This logic can be calculated offchain like

```js
function indices(i, depth) {
  const out = []
  for (let x = 0; x < depth; x++) {
    out.push((i >> x) & 1)
  }
  return out
}
```

I think most root calculation functions should accept just the index of the leaf, and calculate intermediate indices internally.

Reply:
indeed, done: 9e27c98ef556e3bf4c30e7ae716d126f274cd620